BloomFilter.md

```diff
@@ -49,87 +44,185 @@ pushdown even in columns with high cardinality and when space is at a premium.
-Bloom filter techniques.
-
-First, the block Bloom filter algorithm from Putze et al.'s [Cache-, Hash- and Space-Efficient
-Bloom filters](http://algo2.iti.kit.edu/documents/cacheefficientbloomfilters-jea.pdf) is used.
-The block Bloom filter consists of a sequence of small Bloom filters, each of which can fit
-into one cache-line. For best performance, those small Bloom filters are loaded into memory
-cache-line-aligned. For each potential element, the first hash value selects the Bloom filter block
-to be used. Additional hash values are then used to set or test bits as usual, but only inside
-this one block. As for Parquet's initial implementation, each block is 256 bits. When inserting or
-finding value, the first hash of that value is used to index into the array of blocks and pick a
-single one. This single block is then used for the remaining part of the operation.
-
-Second, the remaining part of the operation within the single block uses the folklore split Bloom
-filter technique, as described in section 2.1 of [Network Applications of Bloom Filters:
-A Survey](https://www.eecs.harvard.edu/~michaelm/postscripts/im2005b.pdf). Instead of having one
-array of size `m` shared among all the hash functions, each hash function has a range of `m/k`
-consecutive bit locations disjoint from all the others. The total number of bits is still
-`m`, but the bits are divided equally among the `k` hash functions. This approach
-can be useful for implementation reasons, for example, dividing the bits among the hash functions
-may make parallelization of array accesses easier and take utilization of SSE. As for Parquet's
-implementation, it divides the 256 bits in each block up into eight contiguous 32-bit lanes and
-sets or checks one bit in each lane.
-
-#### Algorithm
-In the initial algorithm, the most significant 32 bits from the hash value are used as the
-index to select a block from bitset. The lower 32 bits of the hash value, along with eight
-constant salt values, are used to compute the bit to set in each lane of the block.
-The salt values are used to construct following different hash functions as described in
-[Multiplicative hashing](https://en.wikipedia.org/wiki/Hash_function#Multiplicative_hashing):
-
-hash<sub>i</sub>(x) = salt<sub>i</sub> * x >> y
-
-Since the target hash value is `[0, 31]`, so we right shift `y = 27` bits. As a result, eight
-hash values, which are indexes of the tiny bloom filter, are generated.
-
-```c
-// 8 SALT values used to compute bit pattern
-static const uint32_t SALT[8] = {0x47b6137bU, 0x44974d91U, 0x8824ad5bU, 0xa2b7289dU,
-  0x705495c7U, 0x2df1424bU, 0x9efc4947U, 0x5c6bfb31U};
-
-// key: the lower 32 bits of hash result
-// mask: the output bit pattern for a tiny Bloom filter
-void Mask(uint32_t key, uint32_t mask[8]) {
-  for (int i = 0; i < 8; ++i) {
-    mask[i] = key * SALT[i];
+
+The section describes split block Bloom filters, which is the first
```

Comment:
Do we need a header for the section here?

Reply:
I don't think so, since this is the only BF algorithm in the spec right now.